src/stress/adapter/device_allocation.spec.ts

```diff
@@ -0,0 +1,264 @@
+    const kMemoryBlockData = new Uint8Array(kMemoryBlockSize);
+
+    const device: GPUDevice = await adapter.requestDevice();
+    const buffers = [];
+    for (let memory = 0; memory < kTotalMemorySize; memory += kMemoryBlockSize) {
+      const buffer = device.createBuffer({
+        size: kMemoryBlockSize,
+        usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_DST,
+      });
+
+      // Write out to the buffer to make sure that it has backing memory.
+      device.queue.writeBuffer(buffer, 0, kMemoryBlockData, 0, kMemoryBlockData.length);
+      buffers.push(buffer);
+    }
+    return { device, objects: buffers };
+  }
+}
+
+export const g = makeTestGroup(DeviceAllocationTests);
+
+g.test('coexisting')
+  .desc(`Tests allocation of many coexisting GPUDevice objects.`)
+  .params(u => u.combine('adapterType', kAdapterTypes))
+  .fn(async t => {
+    const { adapterType } = t.params;
+    const adapter = await getGPU().requestAdapter(kAdapterTypeOptions[adapterType]);
+    assert(adapter !== null, 'Failed to get adapter.');
+
+    // Based on Vulkan comformance test requirement to be able to create multiple devices.
+    const kNumDevices = 5;
+
+    const devices = [];
+    for (let i = 0; i < kNumDevices; ++i) {
+      const device: GPUDevice = await adapter.requestDevice();
+      devices.push(device);
+    }
+  });
+
+g.test('continuous,with_destroy')
+  .desc(
+    `Tests allocation and destruction of many GPUDevice objects over time. Device objects
+are sequentially requested with a series of device allocated objects created on each
+device. The devices are then destroyed to verify that the device and the device allocated
+objects are recycled over a very large number of iterations.`
+  )
+  .params(u => u.combine('adapterType', kAdapterTypes))
+  .fn(async t => {
+    const { adapterType } = t.params;
+    const adapter = await getGPU().requestAdapter(kAdapterTypeOptions[adapterType]);
+    assert(adapter !== null, 'Failed to get adapter.');
+
+    // Since devices are being destroyed, we should be able to create many devices.
+    const kNumDevices = 100;
+    const kFunctions = [
+      t.createDeviceAndBuffers,
+      t.createDeviceAndComputeCommands,
+      t.createDeviceAndRenderCommands,
+    ];
+
+    const deviceList = [];
+    const objectLists = [];
+    for (let i = 0; i < kNumDevices; ++i) {
+      const { device, objects } = await kFunctions[i % kFunctions.length](adapter);
+      t.expect(objects.length > 0, 'unable to allocate any objects');
+      deviceList.push(device);
+      objectLists.push(objects);
+      device.destroy();
+    }
+  });
+
+g.test('continuous,no_destroy')
+  .desc(
+    `Tests allocation and implicit GC of many GPUDevice objects over time. Objects are
+sequentially requested and dropped for GC over a very large number of iterations. Note
+that without destroy, we do not create device allocated objects because that will
+implicitly keep the device in scope.`
```

Comment:
is this true? if we just do: 
```
await kFunctions[i % kFunctions.length](adapter);
```

without assigning to anything on the left hand side, they should all still be GC'ed

Reply:
Yea, but in the previous test, we are purposely keeping track of all the objects which acts as a test control to ensure that the device destroy is the one recycling memory. In this case, what I wanted to do was only keep track of the objects, but not the device and test the same behavior. Unfortunately, though, that doesn't work because by keeping the objects, the device will be kept alive and hence won't be GC-ed.

If we don't assign anything to the left side, it's not guaranteed that the recycling of resources is due to the device being GC-ed since the objects are being GC-ed as well. (I could still include that, but I'm not sure what value it brings atm, assuming we are going to eventually implement the other destroy stress tests for the sub-device objects.)

Reply:
Resolved offline.